assembly/src/assembler/mod.rs

```diff
@@ -22,6 +23,98 @@ mod tests;
 
 type ProcedureCache = BTreeMap<ProcedureId, Procedure>;
 
+// MODULE PROVIDER
+// ================================================================================================
+
+/// A module provider that will store the procedure ASTs used during the compilation process.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ModuleProvider {
```

Comment:
Let's move this struct into its own file - e.g., `/src/assembler/module_provider.rs`.

Also, I wonder if we should track a set of library root namespaces added to the provider. This way, if two libraries with the same root namespace get added two or more times, we can give a more informative error.

Reply:
Done; this will simplify the assembler root